runtime/wasm/src/host_exports.rs

```diff
@@ -335,6 +343,53 @@ where
         )
     }
 
+    // Read the IPFS file `link`, split it into JSON objects, and invoke
+    // the exported function `callback` on each JSON object. The successful
+    // return value contains all entity operations that were produced by the
+    // callback invocations. Each invocation of `callback` happens in its own
+    // instance of a WASM module, which is identical to `module` when it was
+    // first started.
+    pub(crate) fn ipfs_map(
+        &self,
+        module: &WasmiModule<E, L, S, U>,
+        link: String,
+        callback: &str,
+        flags: Vec<String>,
+    ) -> Result<Vec<EntityOperation>, HostExportError<impl ExportError>> {
+        const JSON_FLAG: &str = "json";
+        if !flags.contains(&JSON_FLAG.to_string()) {
+            return Err(HostExportError(format!("Flags must contain 'json'")));
+        }
+
+        let valid_module = module.valid_module.clone();
+        let ctx = module.ctx.clone();
+        let callback = callback.to_owned();
+        // Create a base error message to avoid borrowing headaches
+        let errmsg = format!(
+            "ipfs_map: callback '{}' failed when processing file '{}'",
+            &*callback, &link
+        );
+
+        let operations = self.block_on(
+            self.link_resolver
+                .json_stream(&Link { link })
+                .and_then(|stream| {
+                    stream
+                        .and_then(move |v| {
```

Comment:
I have another request here: Could we log some progress here every now and then? Either every 1000th line or so (with an RwLock counter or similar). Maybe the stream could even emit stats along with the value, like bytes read and lines read that we can use.

Reply:
I've fixed things up so that we log progress every 15s on a logger dedicated to this ipfs_map invocation; the log message is
```
                               debug!(
                                    logger,
                                    "processed {} lines in {}ms so far",
                                    sv.line,
                                    start.elapsed().as_millis()
                                );
```

Reply:
Actually, I'll log it in seconds; millis is silly here if we only log every 15s

Reply:
Awesome, I like it!